src/ttsim/interface_dag_elements/specialized_environment_from_policy_inputs.py

```diff
@@ -0,0 +1,224 @@
+    ParamFunction,
+    PolicyFunction,
+    PolicyInput,
+    param_function,
+    policy_function,
+)
+from ttsim.tt.param_objects import ParamObject
+
+if TYPE_CHECKING:
+    import datetime
+    from collections.abc import Callable
+    from types import ModuleType
+    from typing import Any
+
+    import networkx as nx
+
+    from ttsim.typing import (
+        OrderedQNames,
+        PolicyEnvironment,
+        QNameData,
+        SpecEnvWithoutTreeLogicAndWithDerivedFunctions,
+        SpecEnvWithPartialledParamsAndScalars,
+        SpecEnvWithProcessedParamsAndScalars,
+        UnorderedQNames,
+    )
+
+
+@interface_function()
+def qnames_to_derive_functions_from(
+    labels__input_columns: UnorderedQNames,
+    labels__policy_inputs: UnorderedQNames,
+    labels__grouping_levels: OrderedQNames,
+) -> UnorderedQNames:
+    """The qnames to derive functions from.
+
+    Derived functions should be created based on the actual input columns, and the
+    policy inputs (if their base name is not already in the input columns).
+    """
+    pattern_all = get_re_pattern_for_all_time_units_and_groupings(
+        time_units=list(TIME_UNIT_LABELS),
+        grouping_levels=labels__grouping_levels,
+    )
+    base_names_input_columns = {
+        pattern_all.fullmatch(qn).group("base_name") for qn in labels__input_columns
+    }
+
+    out = set(labels__input_columns)
+    for pi in labels__policy_inputs:
+        match = pattern_all.fullmatch(pi)
+        base_name = match.group("base_name")
+        if base_name in base_names_input_columns:
+            continue
+        out.add(pi)
+    return out
+
+
+@interface_function()
+def without_tree_logic_and_with_derived_functions(
+    policy_environment: PolicyEnvironment,
+    qnames_to_derive_functions_from: UnorderedQNames,
+    labels__column_targets: OrderedQNames,
+    labels__param_targets: OrderedQNames,
+    labels__top_level_namespace: UnorderedQNames,
+    labels__grouping_levels: OrderedQNames,
+) -> SpecEnvWithoutTreeLogicAndWithDerivedFunctions:
+    """Return a flat policy environment with derived functions.
+
+    Two steps:
+    1. Remove all tree logic from the policy environment.
+    2. Add derived functions to the policy environment.
+
+    """
+    qname_env_without_tree_logic = _remove_tree_logic_from_policy_environment(
+        qname_env=dt.flatten_to_qnames(policy_environment),
+        labels__top_level_namespace=labels__top_level_namespace,
+    )
+    return _add_derived_functions(
+        qname_env_without_tree_logic=qname_env_without_tree_logic,
+        tt_targets=labels__column_targets + labels__param_targets,
+        input_columns=qnames_to_derive_functions_from,
+        grouping_levels=labels__grouping_levels,
+    )
+
+
+@interface_function()
+def without_processed_data_nodes_with_dummy_callables(
+    without_tree_logic_and_with_derived_functions: SpecEnvWithoutTreeLogicAndWithDerivedFunctions,  # noqa: E501
+    processed_data: QNameData,
```

Comment:
I think this module should only ever depend on `input_data__columns`. If calling the version from `specialized_environment`, just create a dummy data structure from that?